### PR TITLE
Proper handling sandbox options

### PR DESCRIPTION
**Pull request by fcrisciani**
- Restructure common to pkg directory to be more compliant with the golang directory structure
- Use the resolv.conf passed from the daemon

Addresses: https://github.com/docker/libnetwork/issues/2187
Moby side: https://github.com/moby/moby/pull/37485

### CODE REVIEW

**Comment by ctelfer** [comment]
Hey.  Definitely a change that Docker engine needs.   But not sure on the details, fwiw.

First:  I'm 50/50 on splitting out the utility libraries in "common/" to separate sub-packages.   Of course I'm not really clear on why "common/" is split out as it is.  What makes "caller", "setmatrix" and now "procfs" special?  Maybe they should all just go to the top level?  Or maybe other stand-alones should go in "internal/"?  Hrmm... 

Second:  you have a procfs_linux.go but no procfs_unimplemented.go.  Probably need that.

Third: is checking for a particular binary running the right way to do this?  K8s is mainly doing it to issue a warning so that's "safe" in a sense.  But transparently substituting in libnetwork doesn't seem robust.  Maybe we should do what k8s is doing even more so:  require the user to provide an option for a default resolv.conf file when starting up the engine.  Libnetwork does have (sandbox) options to specify the source of the resolv.conf file already, so this wouldn't necessarily mean a libnetwork change at all.  It could just be a moby change.  (We could still do the warning if systemd-resolverd is detected and the resolv.conf option isn't overridden.)

**Comment by fcrisciani** [comment]
replying to the concerns:
1) the 3 utilities are not sharing anything that reason them to be within the same package. The folder internal holds also the meaning of disallowing the use of the library outside of this project.

2) that only if the procfs code is used by files that are, in this case, compiled also for non linux OS. Compilations succeed because in this commit procfs is already being called by a file named `_linux`

3) Will have to check such option that you are mentioning. For what I can see there, the resolv.conf is hardcoded in the get, that's why I added there the fix.

**Comment by fcrisciani** [comment]
but I definitely see how undesirable is doing it for each sandbox creation

File: internal/procfs/procfs_linux.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package procfs
+
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (abhi) [inline internal/procfs/procfs_linux.go]: Is the copyright needed?
cc @thaJeztah for advice

**Comment by fcrisciani** [comment]
@abhi @ctelfer can you guys take anther look? on the description I added the moby side of it

File: sandbox_dns_unix.go [hunk 2]
```diff
@@ -81,7 +81,7 @@ func (sb *sandbox) buildHostsFile() error {
 	}
 
 	// This is for the host mode networking
-	if sb.config.originHostsPath != "" {
+	if sb.config.useDefaultSandBox && len(sb.config.extraHosts) == 0 {
 		if err := copyFile(sb.config.originHostsPath, sb.config.hostsPath); err != nil && !os.IsNotExist(err) {
```
Comment (ctelfer) [inline sandbox_dns_unix.go]: nit -- assumes that if the condition holds true that originHostsPath has also been set.  This is true in the moby patch of course, but is unclear from just reading the libnetwork code.  Suggest at least putting a comment on OptionUseDefaultSandbox() to say that one must also add OptionOriginHostsPath() or one or more OptionExtraHost() options or expect an error.  Same for OptionOriginResolvConfPath() and OptionDNS()/OptionDNSSearch()/OptionDNSOptions()...

**Comment by fcrisciani** [comment]
@ctelfer added comments requested, feel free to merge if they look ok

**Pull request merged**